### PR TITLE
stoic21 use lightning instead of manager

### PR DESCRIPTION
**Pull request by mosheraboh**
runner_stoic21.py is the main file to review when moving to use PyTorch lightning-based trainer.
All the rest are minors.

### CODE REVIEW

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py [hunk 1]
```diff
@@ -91,30 +96,42 @@
 TRAIN_COMMON_PARAMS['data.train_folds'] = [0, 1, 2, 3]
 TRAIN_COMMON_PARAMS['data.validation_folds'] = [4]
 
+# ===============
+# PL Trainer
+# ===============
+TRAIN_COMMON_PARAMS['trainer.num_epochs'] = 50
+TRAIN_COMMON_PARAMS['trainer.num_devices'] = 1
```
Comment (michalozeryflato) [inline examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:103]: trainer.num_devices should be as NUM_GPUS

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py [hunk 2]
```diff
@@ -285,21 +292,24 @@ def run_infer(paths: dict, infer_common_params: dict):
         infer_sample_ids += folds[fold]
 
     infer_dataset = STOIC21.dataset(paths["data_dir"], paths["cache_dir"], sample_ids=infer_sample_ids, train=False)
+    infer_dataloader = DataLoader(dataset=infer_dataset, collate_fn=CollateDefault(), batch_size=2, num_workers=2)
```
Comment (SagiPolaczek) [inline examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:295]: accidentally not using the infer_common_params's batch_size and num_workers?

**Review by SagiPolaczek — COMMENTED** [review]
Thanks for the reference Moshiko! I updated the ISIC and I'll soon open a PR.

I left some comments for small things that I found during the review :)

Another thing that a bit bothered me is that we refer the `TRAIN_COMMON_PARAMS` in `run_train` as `train_params` while refer the INFER and EVAL in the `run_infer` and `run_eval` as `infer_common_params` and `eval_common_params`.  Nothing too serious but a bit ugly and since we soon going public it might be nice change it 😄  btw In the walkthrough_template.py file it all with 'common'.

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py [hunk 3]
```diff
@@ -245,7 +251,7 @@ def run_infer(paths: dict, infer_common_params: dict):
 
     # convert list of batch outputs into a dataframe
     infer_df = convert_predictions_to_dataframe(predictions)
-    save_dataframe(infer_df, infer_common_params['infer_filename'])
+    save_dataframe(infer_df, infer_file)
     
 
 ######################################
```
Comment (SagiPolaczek) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist.py:257]: should be Eval Common Params

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py [hunk 4]
```diff
@@ -245,7 +251,7 @@ def run_infer(paths: dict, infer_common_params: dict):
```
Comment (SagiPolaczek) [inline examples/fuse_examples/imaging/classification/mnist/run_mnist.py:251]: in lines 247, 248 we are using a global ver `TRAIN_COMMON_PARAMS` which works but I think we better use the local `infer_common_params` (values should pass between dicts first)

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py [hunk 5]
```diff
@@ -207,51 +205,54 @@ def run_train(paths: dict, train_params: dict):
...
-    # =====================================================================================
-    #  Callbacks
-    # =====================================================================================
-    callbacks = [
-        # default callbacks
-        TensorboardCallback(model_dir=paths['model_dir']),  # save statistics for tensorboard
-        MetricStatisticsCallback(output_path=paths['model_dir'] + "/metrics.csv"),  # save statistics a csv file
-        TimeStatisticsCallback(num_epochs=train_params['manager.train_params']['num_epochs'], load_expected_part=0.1)  # time profiler
-    ]
+    validation_metrics = copy.deepcopy(train_metrics) # use the same metrics in validation as well
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:212]: Yes, PyTorch lightning run train epoch, validation epoch and then evaluate.
If you won't do it you'll mix train and validation which will lead to an error

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py (hunk 4 above)
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist.py:251]: Already fixed it and missed a push, thanks

File: examples/fuse_examples/imaging/classification/mnist/run_mnist.py (hunk 3 above)
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/mnist/run_mnist.py:257]: Already fixed it and missed a push, thanks

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py [hunk 6]
```diff
@@ -207,51 +205,54 @@ def run_train(paths: dict, train_params: dict):
...
+ 
+    # either a dict with arguments to pass to ModelCheckpoint or list dicts for multiple ModelCheckpoint callbacks (to monitor and save checkpoints for more then one metric). 
+    best_epoch_source = dict(
+        monitor="validation.metrics.auc",
+        mode="max",
+    )
+    
+    # create optimizer
+    optimizer = optim.SGD(model.parameters(), lr=train_common_params['opt.lr'], weight_decay=train_common_params['opt.weight_decay'], momentum=0.9, nesterov=True)
```
Comment (SagiPolaczek) [inline examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:221]: Just a minor comment regarding the documentation:
We split the code in `run_train` into sections using (for example):
```
# ==========
# Model
# ==========
```

Now I noticed that the optimizer and lr_sched are assigned to `#  Metrics` section. 
Do you think we should add another one section between `# Metrics` and `# Train` ? (Also for the other examples and the walkthrough_template)

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py [hunk 7]
```diff
@@ -312,6 +322,8 @@ def run_infer(paths: dict, infer_common_params: dict):
 # Eval Template
 ######################################
 def run_eval(paths: dict, eval_common_params: dict):
+    infer_file = os.path.join(paths['inference_dir'], eval_common_params['infer_filename'])
```
Comment (SagiPolaczek) [inline examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:325]: `infer_file` is being defined twice (also in `run_infer`), we can consider moving it to `PATHS`.

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py (hunk 2 above)
Comment (SagiPolaczek) [reply examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:295]: @mosheraboh so it was on purpose? :)

**Review by SagiPolaczek — APPROVED** [review]
Looks good :)
I added some minor comments

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py (hunk 7 above)
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:325]: You might have few inference files so I prefer to keep them separated

File: examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py (hunk 6 above)
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/stoic21/runner_stoic21.py:221]: Fixed it, thanks

**Pull request merged**